plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/logging/KotlinPlaceholderCountMatchesArgumentCountInspection.kt

```diff
@@ -0,0 +1,208 @@
+// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+
+package org.jetbrains.kotlin.idea.inspections.logging
+
+import com.intellij.codeInspection.ProblemsHolder
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.builtins.StandardNames.FqNames.throwable
+import org.jetbrains.kotlin.idea.KotlinBundle
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.idea.caches.resolve.resolveToCall
+import org.jetbrains.kotlin.idea.core.resolveType
+import org.jetbrains.kotlin.idea.inspections.AbstractKotlinInspection
+import org.jetbrains.kotlin.idea.intentions.receiverType
+import org.jetbrains.kotlin.idea.refactoring.fqName.fqName
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtValueArgument
+import org.jetbrains.kotlin.psi.callExpressionVisitor
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.constants.evaluate.ConstantExpressionEvaluator
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.typeUtil.supertypes
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+class KotlinPlaceholderCountMatchesArgumentCountInspection : AbstractKotlinInspection() {
+
+    enum class LoggerType {
+        SLF4J_LOGGER, LOG4J_LOGGER, LOG4J_BUILDER
+    }
+
+    companion object {
+        private val LOGGING_METHODS = setOf("log", "trace", "debug", "info", "warn", "error", "fatal")
+        private val LOGGER_CLASSES =
+            mapOf(
+                "org.slf4j.Logger" to LoggerType.SLF4J_LOGGER,
+                "org.apache.logging.log4j.Logger" to LoggerType.LOG4J_LOGGER,
+                "org.apache.logging.log4j.LogBuilder" to LoggerType.LOG4J_BUILDER
+            )
+    }
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) = callExpressionVisitor(fun(call) {
+        val name = call.calleeExpression?.text
+        if (name == null || !LOGGING_METHODS.contains(name)) {
+            return
+        }
+
+        val method = call.resolveToCall()?.resultingDescriptor ?: return
+        val parents = method.receiverType().findParents(LOGGER_CLASSES.keys)
+        if (parents.isEmpty()) {
+            return
+        }
+        val loggerType = LOGGER_CLASSES[parents.first()] ?: return
+
+        val valueParameters = method.valueParameters
+        if (valueParameters.isEmpty()) {
+            return
+        }
+        val indexString: Int
+        if (KotlinBuiltIns.isString(valueParameters[0]?.type)) {
+            indexString = 0
+        } else {
+            if (valueParameters.size < 2 || !KotlinBuiltIns.isString(valueParameters[1]?.type)) {
```

Comment:
Without a context these code is difficult to understand. I had to peek at the slf4j documentation.
Can I ask you to add a comment clarifying what happens with this code snippet (there are signatures..., pattern parameter can be here or there, etc), please.

Reply:
Add some examples - could you have a look if they are clear?

Reply:
Looks much better to me, thanks